File: Bouncer/src/shared/platforms.ts
// Platform registry — single source of truth for everything Bouncer needs
// to know about a supported site. There is exactly ONE source: the JSON
// file `platforms.config.json`. Adding a platform (or changing which builds
// it ships on) is a single edit there. Every consumer reads from it:
//   - this TS registry (content script, popup, background pipeline)
//   - the build-time manifest generator (generate-manifests.mjs)
//   - the iOS app's native picker / WebView injection (Platforms.swift,
//     which decodes the same bundled JSON)
//
// A platform is present on a given build only when its `targets` array
// includes that build's target. `PLATFORMS` below is already filtered to
// the current build target, so a platform absent from this build never
// appears in any UI, storage key, or pipeline loop — no partial state.

import platformsConfig from './platforms.config.json';

/** Literal-tuple of every supported platform id. SiteId is the union of
 *  ids the codebase understands; platforms.config.json supplies a subset of
 *  these (whichever are actually shipped) along with their runtime data. */
export const PLATFORM_IDS = ['twitter', 'youtube', 'linkedin'] as const;

export type SiteId = typeof PLATFORM_IDS[number];

/** Build target a platform can ship on. Mirrors build.js `--target=` and
 *  generate-manifests.mjs. `safari` covers the iOS + macOS apps. */
export type BuildTarget = 'chrome' | 'firefox' | 'safari';

/** The complete, JSON-serializable record for one platform. This is the
 *  shape stored in platforms.config.json and shared verbatim by the TS
 *  bundle, the manifest generator, and (decoded) the iOS app. Everything a
 *  platform needs lives here so no consumer has to hardcode it. */
export interface PlatformConfig {
  readonly id: SiteId;
  /** Human-facing display name (popup labels, picker rows). */
  readonly displayName: string;
  /** Builds this platform ships on. Filters host_permissions /
   *  content_scripts (manifest), `PLATFORMS` (this file), and the iOS
   *  picker (Platforms.swift). */
  readonly targets: readonly BuildTarget[];
  /** Pattern used in host_permissions, content_scripts, and
   *  web_accessible_resources matches. */
  readonly manifestHost: string;
  /** Path to the bundled adapter JS, relative to the extension root. */
  readonly adapterScript: string;
  /** Path to the platform's stylesheet, relative to the extension root. */
  readonly cssPath: string;
  /** Additional web-accessible files this platform's adapter loads via
   *  chrome.runtime.getURL (page-world helper scripts, etc.). */
  readonly extraWebAccessible: readonly string[];
  /** Where the platform's feed lives. */
  readonly feedUrl: string;
  /** Alternate URL for first-launch when the user is not signed in (X has a
   *  login flow; LinkedIn/YouTube don't gate this way). `null` when unused. */
  readonly loginUrl: string | null;
  /** Every host this platform owns — the primary site plus related CDN /
   *  API domains. Root + subdomain matches derive `hostPattern`, and the
   *  iOS WebView's navigation guard flat-maps these into its allowlist. */
  readonly hostRoots: readonly string[];
}

/** Build the runtime host-matching RegExp from `hostRoots`: matches each
 *  root exactly or as a subdomain (`x.com` and `mobile.x.com`, not
 *  `notx.com`). Mirrors Platforms.swift `matches(host:)`. */
function hostPatternFor(hostRoots: readonly string[]): RegExp {
  const alternation = hostRoots
    .map(root => root.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'))
    .join('|');
  return new RegExp(`(^|\\.)(${alternation})$`, 'i');
}

export type PlatformDef = PlatformConfig & {
  /** Regex matched against `location.hostname`, derived from `hostRoots`. */
  readonly hostPattern: RegExp;
};

/** Current build target. build.js replaces `process.env.BOUNCER_TARGET`
 *  with a literal at bundle time; outside a build (e.g. vitest) it is
 *  undefined and we fall back to `chrome`. */
const CURRENT_TARGET: BuildTarget =
  (process.env.BOUNCER_TARGET as BuildTarget) || 'chrome';

/** Platforms shipped on THIS build, with the derived host pattern attached.
 *  Filtered by `targets`, so a platform not shipping on this target is
 *  invisible to every downstream consumer. */
export const PLATFORMS: readonly PlatformDef[] =
  (platformsConfig as readonly PlatformConfig[])
    .filter(cfg => cfg.targets.includes(CURRENT_TARGET))
    .map(cfg => ({ ...cfg, hostPattern: hostPatternFor(cfg.hostRoots) }));

// ---------------------------------------------------------------------------
// Lookup helpers
// ---------------------------------------------------------------------------

/** Find a platform by its canonical id. Returns undefined for unknown ids. */
export function platformById(id: string): PlatformDef | undefined {
  return PLATFORMS.find(p => p.id === id);
}

/** Find a platform whose host pattern matches the given hostname. */
export function platformFromHost(host: string): PlatformDef | undefined {
  return PLATFORMS.find(p => p.hostPattern.test(host));
}

// ---------------------------------------------------------------------------
// Derived storage-key helpers — every consumer of "{id}Enabled" /
// "descriptions_{id}" goes through these instead of building the strings
// inline, so adding a new platform doesn't require teaching the rest of the
// codebase about a new literal.
// ---------------------------------------------------------------------------

/** chrome.storage.local key for the per-platform master enable toggle. */
export function enabledStorageKey(id: SiteId): `${SiteId}Enabled` {
  return `${id}Enabled`;
}

/** chrome.storage.local key for the per-platform filter-phrase list. */
export function descriptionsStorageKey(id: SiteId): `descriptions_${SiteId}` {
  return `descriptions_${id}`;
}

File: Bouncer_xcode/iOS (App)/Platforms.swift
//
//  Platforms.swift
//  iOS (App)
//
//  Per-platform iOS data (feed URL, adapter/CSS names, host roots) lives in
//  `defined` below. WHICH of those platforms actually appear is gated by the
//  shared Bouncer/src/shared/platforms.config.json (its `targets` must include
//  "safari") — the single source of truth for platform enablement, shared with
//  the extension build. The picker, WebView script/CSS injection,
//  URL-to-platform sync, and feed-URL switch all read `Platforms.all`, which is
//  `defined` intersected with what the config enables.
//

import Foundation

struct PlatformDef {
    /// Canonical id — matches the SiteId values used on the JS side
    /// ("twitter", "youtube", "linkedin"). Stored as String so the existing
    /// FilterSheetViewModel.selectedPlatform: String contract is unchanged.
    let id: String

    /// Human-facing label used in PlatformPickerView's rows.
    let displayName: String

    /// Initial URL the WebView loads when the user picks this platform.
    /// For X this is the home feed if signed in, falling back to the login
    /// flow on first launch — see `loginURL`.
    let feedURL: String

    /// First-launch URL when the user has never signed in. nil for platforms
    /// that handle authentication internally without needing a special entry
    /// point (YouTube, LinkedIn).
    let loginURL: String?

    /// Bundled-resource filename (without .js extension) for this platform's
    /// adapter script in dist/. FilteredWebView injects each adapter as a
    /// WKUserScript at document-end; the adapter self-guards by hostname.
    let adapterScriptName: String

    /// Filename (without .css extension) of the platform's stylesheet.
    let cssFile: String

    /// Bundle subdirectory holding `cssFile`. Matches the desktop manifest's
    /// content_scripts.css paths under adapters/<id>/.
    let cssSubdir: String

    /// True when this platform owns `host` — covers root + subdomain matches.
    /// Used by syncPlatformToCurrentSite() to map the WebView's current host
    /// back to a platform id.
    func matches(host: String) -> Bool {
        let h = host.lowercased()
        return hostRoots.contains { root in h == root || h.hasSuffix("." + root) }
    }

    /// Domains this platform considers part of itself, including any related
    /// CDN / API hosts that should be permitted by `allowedHosts` in
    /// FilteredWebView's navigation guard. Stored as raw domains; the
    /// WebView allowedHosts list flat-maps these from every PlatformDef.
    let hostRoots: [String]
}

enum Platforms {
    /// Per-platform iOS data, in PlatformPickerView row order (top to bottom).
    /// WHICH of these actually appear is gated by `enabledIds` — see `all`.
    private static let defined: [PlatformDef] = [
        PlatformDef(
            id: "twitter",
            displayName: "X (Twitter)",
            feedURL: "https://x.com/home",
            loginURL: "https://x.com/i/flow/login",
            adapterScriptName: "TwitterAdapter",
            cssFile: "twitter",
            cssSubdir: "adapters/twitter",
            hostRoots: [
                "x.com", "twitter.com", "t.co", "twimg.com",
                "pbs.twimg.com", "abs.twimg.com", "video.twimg.com",
            ]
        ),
// Millan todo: add back YT to ios
//         PlatformDef(
//             id: "youtube",
//             displayName: "YouTube",
//             feedURL: "https://www.youtube.com/",
//             loginURL: nil,
//             adapterScriptName: "YouTubeAdapter",
//             cssFile: "youtube",
//             cssSubdir: "adapters/youtube",
//             hostRoots: [
//                 "youtube.com", "m.youtube.com", "youtu.be",
//                 "ytimg.com", "ggpht.com", "googlevideo.com",
//                 "accounts.youtube.com",
//             ]
//         ),
        PlatformDef(
            id: "linkedin",
            displayName: "LinkedIn",
            feedURL: "https://www.linkedin.com/feed/",
            loginURL: nil,
            adapterScriptName: "LinkedInAdapter",
            cssFile: "linkedin",
            cssSubdir: "adapters/linkedin",
            hostRoots: [
                "linkedin.com", "licdn.com",
                "static.licdn.com", "media.licdn.com",
            ]
        ),
        PlatformDef(
            id: "instagram",
            displayName: "Instagram",
            // Bouncer's Instagram surface is the Reels viewer, not the home
            // feed — matches PLATFORM_RUNTIME.instagram.feedUrl on the JS side.
            feedURL: "https://www.instagram.com/reels/",
            loginURL: nil,
            adapterScriptName: "InstagramAdapter",
            cssFile: "instagram",
            cssSubdir: "adapters/instagram",
            hostRoots: [
                "instagram.com", "cdninstagram.com",
                "fbcdn.net", "facebook.com",
            ]
        ),
    ]

    /// Platforms actually shown on iOS: the entries `defined` above that are
    /// also enabled for the "safari" target in the shared
    /// platforms.config.json — the single source of truth for which platforms
    /// ship where. A platform missing from the config (or without "safari" in
    /// its targets) is dropped from the picker, WebView injection, and host
    /// guard, so the native UI can never offer a platform the JS pipeline
    /// doesn't know about.
    static let all: [PlatformDef] = defined.filter { enabledIds.contains($0.id) }

    /// Canonical ids enabled for this app's target ("safari"), read from the
    /// bundled platforms.config.json.
    private static let enabledIds: Set<String> = loadEnabledIds()

    private static func loadEnabledIds() -> Set<String> {
        guard let url = Bundle.main.url(
                forResource: "platforms.config", withExtension: "json", subdirectory: "shared"),
              let data = try? Data(contentsOf: url),
              let entries = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            assertionFailure("platforms.config.json missing or invalid in app bundle")
            return []
        }
        var ids = Set<String>()
        for entry in entries {
            guard let id = entry["id"] as? String,
                  let targets = entry["targets"] as? [String],
                  targets.contains("safari") else { continue }
            ids.insert(id)
        }
        return ids
    }

    /// Lookup by canonical id (e.g., "twitter"). Returns nil for unknown ids.
    static func byId(_ id: String) -> PlatformDef? {
        return all.first { $0.id == id }
    }

    /// Lookup by host — finds which platform "owns" a URL. Useful for the
    /// syncPlatformToCurrentSite() flow that maps a WebView URL change back
    /// to the matching platform id.
    static func fromHost(_ host: String) -> PlatformDef? {
        return all.first { $0.matches(host: host) }
    }

    /// Flat list of every host any platform considers its own — used by the
    /// WebView navigation guard's allowedHosts. Auth / system hosts
    /// (Google sign-in, Apple ID, etc.) are added on top of this in
    /// FilteredWebView since they're shared across platforms.
    static var allHostRoots: [String] {
        return all.flatMap { $0.hostRoots }
    }
}
